neurograph.py

```diff
@@ -0,0 +1,170 @@
+import networkx as nx
+from networkx.algorithms import bipartite
+
+#input: DAG g as an adjacent list representation 
+# b is in a's adj list if a --> b
+#construct transitive closure g' using DFS
+#store a path dict (trans_path), where key = vertice v, value = a dict maps reachable nodes u to a path of v to u
+#store a graph dict trans_dict (transitive closure of g, adjacent list representation)
+# for node v, return a dict maps all reachable nodes u to a path of v to u
+
+def helper(g, trans_path, v):
+    if v in trans_path:
+        return trans_path
+    if len(g[v]) == 0:
+        trans_path[v] = {}
+        return trans_path
+    nbrs = g[v]
+    v_dict = {}
+    for u in nbrs:
+        recur = helper(g, trans_path, u)
+        u_dict = recur[u]
+        for node in u_dict:
+            if node not in v_dict:
+                v_dict[node] = [v] + u_dict[node]
+        v_dict[u] = [v, u]
+    trans_path[v] = v_dict
+    return trans_path
+
+def transitive_closure(g):
+    trans_path = {} 
+    trans_dict = {}
+    for v in g:
+        helper(g, trans_path, v)
+    
+    for v in trans_path:
+        reachables = []
+        reachable_dict = trans_path[v]
+        for u in reachable_dict:
+            reachables.append(u)
+        trans_dict[v] = reachables
+    return trans_path, trans_dict
+
+#idea is the following: 
+# 1. create the bipartite graph. split each node v into two nodes v_top and v_bottom, where v_top connects with all outgoing edges of v, 
+# and v_bottom connects with all incoming edges of v.
+# 2. use max matching in networkx to generate the matching
+# 3. recover the path in the transitive closure graph using the bipartite graph matching
+# reference: https://towardsdatascience.com/solving-minimum-path-cover-on-a-dag-21b16ca11ac0
+
+def max_matching(g):
+    #generate bipartite
+    b = nx.Graph()
+    top_nodes = []
+    bottom_nodes = []
+    #generate nodes
+    for v in g:
+        v_top = str(v) + "t"
+        v_bottom = str(v) + "b"
+        top_nodes.append(v_top)
+        bottom_nodes.append(v_bottom)
+    #generate edges
+    edges = []
+    for v in g:
+        nbrs = g[v]
+        for u in nbrs:
+            v_top = str(v) + "t"
+            u_bottom = str(u) + "b"
+            edges.append([v_top, u_bottom])
+    
+    b.add_nodes_from(top_nodes, bipartite=0)
+    b.add_nodes_from(bottom_nodes, bipartite=1)
+    b.add_edges_from(edges)
+
+    #calculate maximum cardinality matching
+    #matching: a dict maps node to node (both ways, represent edges)
+    matching = bipartite.matching.hopcroft_karp_matching(b, top_nodes)
+
+    #generate paths from matching
+    #ret type: list of list
+    paths = []
+    visited_nodes = set()
+    for v in sorted(g.keys()):
```

Comment:
Aren't the keys of the graph just integers in sequence? If so, do we want to pay an O(n log(n)) penalty here instead of just using a for loop potentially? Or is it unavoidable?

Reply:
You are absolutely right. Thank you for pointing this out!

Also, just wondering, do we need to consider the general case that in the input graph, the vertices are named in strings instead of consecutive numbers?

Reply:
Resolved